### ISSUE TITLE
fuzzMouse assertion failure

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/function-builder#62]
Reported in this Skype conversation by @phet-steele:

[6/7/16, 4:35:48 PM] Steele Dalton: well, are you expecting FB to be stable at this moment, or are you still refactoring :)
[6/7/16, 4:37:10 PM] Chris Malley: I've been doing major work on the table feature, but haven't noticed any problem with what's checked into master.  that doesn't mean there aren't any problems, because I haven't done any kind of thorough regression test.
[6/7/16, 4:37:44 PM] Chris Malley: so yes, i'm still working on a new feature that may be buggy. if you've noticed something, i'd be happy to hear specifics.
[6/7/16, 4:40:41 PM] Steele Dalton: Well I get this very verbose assertion, that from reading it is probably a clue for the remaining plethora of errors
[6/7/16, 4:41:00 PM] Steele Dalton: “Assertion failed: This should not be run in the call tree of updateDisplay(). If you see this, it is likely that either the last updateDisplay() had a thrown error and it is trying to be run again (in which case, investigate that error), OR code was run/triggered from inside an updateDisplay() that has the potential to cause an infinite loop, e.g. CanvasNode paintCanvas() call manipulating another Node, or a bounds listener that Scenery missed.”
[6/7/16, 4:41:58 PM] Chris Malley: That is coming from somewhere deep down in scenery. Platform and steps to reproduce? Any stack trace?
[6/7/16, 4:42:36 PM] Steele Dalton: uhh, fuzzMouse on current master…OS X latest
[6/7/16, 4:42:39 PM] Steele Dalton: chrome
[6/7/16, 4:43:08 PM] Steele Dalton: it’s more like a stack journey than a stack trace
[6/7/16, 4:43:31 PM] Steele Dalton: 

```
window.assertions.assertFunction	@	assert.js:23
ensureNotPainting	@	Display.js?bust=1465338480677:1066
markDirtyDrawable	@	BackboneDrawable.js?bust=1465338480677:211
markDirty	@	Drawable.js?bust=1465338480677:282
markDirtyGroup	@	SVGBlock.js?bust=1465338480677:159
markDirty	@	SVGGroup.js?bust=1465338480677:122
markTransformDirty	@	SVGGroup.js?bust=1465338480677:140
trigger0	@	Events.js?bust=1465338480677:284
onTransformChange	@	Node.js?bust=1465338480677:1839
trigger0	@	Events.js?bust=1465338480677:284
invalidate	@	Transform3.js?bust=1465338480677:106
prependTranslation	@	Transform3.js?bust=1465338480677:143
prependTranslation	@	Node.js?bust=1465338480677:1781
translate	@	Node.js?bust=1465338480677:1470
translate	@	Node.js?bust=1465338480677:1480
Node.(anonymous function)	@	Node.js?bust=1465338480677:4249
translateNode	@	MovableNode.js?bust=1465338480677:36
(anonymous function)	@	MovableNode.js?bust=1465338480677:50
emit2	@	Emitter.js?bust=1465338480677:139
_notifyObservers	@	Property.js?bust=1465338480677:129
_setAndNotifyObservers	@	Property.js?bust=1465338480677:117
set	@	Property.js?bust=1465338480677:85
moveTo	@	Movable.js?bust=1465338480677:66
translateMovable	@	MovableNode.js?bust=1465338480677:33
drag	@	MovableNode.js?bust=1465338480677:77
move	@	SimpleDragHandler.js?bust=1465338480677:124
dispatchToPointer	@	Input.js?bust=1465338480677:811
dispatchEvent	@	Input.js?bust=1465338480677:784
branchChangeEvents	@	Input.js?bust=1465338480677:717
moveEvent	@	Input.js?bust=1465338480677:678
mouseMove	@	Input.js?bust=1465338480677:334
fuzzMouseEvents	@	Display.js?bust=1465338480677:1127
stepSimulation	@	Sim.js?bust=1465338480677:690
runAnimationLoop	@	Sim.js?bust=1465338480677:664
```

[6/7/16, 4:43:41 PM] Steele Dalton: haha
[6/7/16, 4:44:24 PM] Chris Malley: are you sure that you have master of everything checked out? because i'm not seeing it.
[6/7/16, 4:44:47 PM] Chris Malley: status.sh should show everything on master.
[6/7/16, 4:44:49 PM] Steele Dalton: the only commit i’m missing should be the one titled TODO
[6/7/16, 4:45:13 PM] Steele Dalton: from 15 minutes ago
[6/7/16, 4:45:18 PM] Chris Malley: you're using master of all dependencies?
[6/7/16, 4:46:24 PM] Steele Dalton: should be, it’s just phettest with every repo up to date…I’ll take a sec to try again
[6/7/16, 4:49:47 PM] Steele Dalton: yup still happens
[6/7/16, 4:50:29 PM] Steele Dalton: everything in chipper’s active-repos is up to date
[6/7/16, 4:51:03 PM] Steele Dalton: the error it runs into first is:

```assertion failed: Bad Image.initialWidth```

[6/7/16, 4:51:18 PM] Steele Dalton: it’s sounding like this is unexpected enough to just become an issue
[6/7/16, 4:51:32 PM] Chris Malley: I'm running with Chrome Version 50.0.2661.102 (64-bit).  When I go to "About", looks like it's now attempting to update.
[6/7/16, 4:51:38 PM] Chris Malley: Your version?
[6/7/16, 4:52:01 PM] Steele Dalton: 51.0.2704.79 (64-bit), but also updating
[6/7/16, 4:53:52 PM] Steele Dalton: still a thing after updating
[6/7/16, 4:54:08 PM] Chris Malley: You were on v51, I was on v50.
[6/7/16, 4:54:41 PM] Steele Dalton: I’m on this now, you? 51.0.2704.84 (64-bit)
[6/7/16, 4:54:44 PM] Chris Malley: Now I'm on Version 51.0.2704.84 (64-bit)
[6/7/16, 4:54:59 PM] Steele Dalton: sweet, try running on phettest with fuzzMouse
[6/7/16, 4:56:39 PM] Steele Dalton: shouldn’t have to pull, since i just did a few minutes ago
[6/7/16, 4:57:17 PM] Chris Malley: are you running "Built Version" for function-builder?
[6/7/16, 4:57:27 PM] Steele Dalton: no, just click the sims name
[6/7/16, 4:58:48 PM] Chris Malley: and then you're manually adding fuzzMouse?
[6/7/16, 4:58:52 PM] Steele Dalton: yea
[6/7/16, 4:59:50 PM] Chris Malley: no problem yet...
[6/7/16, 5:00:20 PM] Chris Malley: fuzzMouse with no value?
[6/7/16, 5:00:30 PM] Steele Dalton: you could add =100
[6/7/16, 5:00:35 PM] Chris Malley: what are you doing?
[6/7/16, 5:00:39 PM] Steele Dalton: either
[6/7/16, 5:00:45 PM] Steele Dalton: seen both
[6/7/16, 5:01:20 PM] Chris Malley: still nothing.
[6/7/16, 5:01:31 PM] Chris Malley: have you tried clearing your cache?
[6/7/16, 5:01:38 PM] Steele Dalton: always
[6/7/16, 5:01:57 PM] Chris Malley: does it always occur quickly?
[6/7/16, 5:02:05 PM] Steele Dalton: nah
[6/7/16, 5:02:07 PM] Chris Malley: ... soon after starting.
[6/7/16, 5:02:18 PM] Chris Malley: how long? or just randomly?
[6/7/16, 5:02:27 PM] Steele Dalton: seems pretty random
[6/7/16, 5:02:51 PM] Chris Malley: you're keeping the other query parameters? (?ea&brand=phet)
[6/7/16, 5:02:58 PM] Steele Dalton: you bet
[6/7/16, 5:03:13 PM] Chris Malley: give me a URL.
[6/7/16, 5:03:31 PM] Steele Dalton: ye of little faith
http://phettest.colorado.edu/function-builder/function-builder_en.html?ea&brand=phet&fuzzMouse=100
[6/7/16, 5:03:45 PM] Chris Malley: i have faith. just want to make sure we're testing the exact same thing.
[6/7/16, 5:04:59 PM] Chris Malley: still nothing.
[6/7/16, 5:05:08 PM] Chris Malley: OS X 10.11.5 ?
[6/7/16, 5:05:12 PM] Steele Dalton: yup

**Comment by pixelzoom** [comment]
function-builder master was at sha @6874d87253cdc75305e108579c1f1968810a63b9 when I was testing this.

**Comment by phet-steele** [comment]
Occurring as well on my Win 10 Chrome.

**Comment by phet-steele** [comment]
Stack journey for the other error above:
```
Uncaught Error: Assertion failed: Bad Image.initialWidth
window.assertions.assertFunction @ assert.js:23
getImageWidth @ Image.js?bust=1465342863850:533
invalidateImage @ Image.js?bust=1465342863850:176
setImage @ Image.js?bust=1465342863850:256
set image @ Image.js?bust=1465342863850:260
updateContent @ ImageCardNode.js?bust=1465342863850:62
(anonymous function) @ CardNode.js?bust=1465342863850:236
emit2 @ Emitter.js?bust=1465342863850:139
_notifyObservers @ Property.js?bust=1465342863850:129
_setAndNotifyObservers @ Property.js?bust=1465342863850:117
set @ Property.js?bust=1465342863850:85
listener @ DerivedProperty.js?bust=1465342863850:69
emit2 @ Emitter.js?bust=1465342863850:139
_notifyObservers @ Property.js?bust=1465342863850:129
_setAndNotifyObservers @ Property.js?bust=1465342863850:117
set @ Property.js?bust=1465342863850:85
moveTo @ Movable.js?bust=1465342863850:66
addNode @ MovableContainer.js?bust=1465342863850:83
moveToInputCarousel @ CardNode.js?bust=1465342863850:310
(anonymous function) @ OutputCardsCarousel.js?bust=1465342863850:77
(anonymous function) @ OutputCardsCarousel.js?bust=1465342863850:75
erase @ OutputCardsCarousel.js?bust=1465342863850:74
reset @ OutputCardsCarousel.js?bust=1465342863850:64
SceneNode._reset @ SceneNode.js?bust=1465342863850:264
reset @ SceneNode.js?bust=1465342863850:387
resetAll @ PatternsScreenView.js?bust=1465342863850:92
(anonymous function) @ PushButtonModel.js?bust=1465342863850:130
fire @ PushButtonModel.js?bust=1465342863850:129
(anonymous function) @ PushButtonModel.js?bust=1465342863850:73
onValueObserver @ Property.js?bust=1465342863850:315
emit2 @ Emitter.js?bust=1465342863850:139
_notifyObservers @ Property.js?bust=1465342863850:129
_setAndNotifyObservers @ Property.js?bust=1465342863850:117
set @ Property.js?bust=1465342863850:85
set @ PropertySet.js?bust=1465342863850:141
up @ ButtonListener.js?bust=1465342863850:48
buttonUp @ DownUpListener.js?bust=1465342863850:123
up @ DownUpListener.js?bust=1465342863850:56
dispatchToPointer @ Input.js?bust=1465342863850:811
dispatchEvent @ Input.js?bust=1465342863850:784
upEvent @ Input.js?bust=1465342863850:654
mouseUp @ Input.js?bust=1465342863850:326
fuzzMouseEvents @ Display.js?bust=1465342863850:1099
stepSimulation @ Sim.js?bust=1465342863850:690
runAnimationLoop @ Sim.js?bust=1465342863850:664
```

**Comment by jonathanolson** [comment]
The following in ImageCardNode:
```js
    updateContent: function( builder, numberOfFunctionsToApply ) {

      if ( numberOfFunctionsToApply === 0 ) {

        // performance optimization
        this.imageNode.image = this.card.image;
      }
      else {

        // {HTMLCanvasElement} run the input image through the builder
        var canvas = builder.applyFunctions( this.card.canvas, numberOfFunctionsToApply );

        // display the output image
        this.imageNode.setImageWithSize( canvas.toDataURL(), canvas.width, canvas.height );
      }
      this.imageNode.center = this.backgroundNode.center;
    }
```

It looks like it's run at least two times, first taking the second branch (sets image and initial dimensions), and then taking the first branch (but with an image of a different width). This causes it to error out, as it detects a different dimension.

I'd recommend setImageWidthSize for all usages if that's reasonable?

The errors afterwards are more of "Scenery stuff was interrupted", and only happen because the assertion interrupts execution.

**Comment by pixelzoom** [comment]
> I'd recommend setImageWidthSize for all usages if that's reasonable?

Thanks @jonathanolson, I'll try that.  Still perplexed about why I'm unable to reproduce it with the same configuration as @phet-steele.

**Comment by pixelzoom** [comment]
@jonathanolson a question... Is there any performance difference between these 2 methods of setting the image?

```
this.imageNode.setImageWithSize( image, image.width, image.height );
this.imageNode.setImageWithSize( canvas.toDataURL(), canvas.width, canvas.height );
```

The only reason that I wasn't using `setImageWithSize( canvas.toDataURL(),..)` in all cases was that canvas operations are needed while a card is still in the input carousel, and I suspected that `setImageWithSize( image,...)` might be more efficient.  But if that's not the case, then I'll use `setImageWithSize( canvas.toDataURL(),..)`  for all cases.

**Comment by pixelzoom** [comment]
@phet-steele Fix pushed, please pull master and verify.

**Comment by pixelzoom** [comment]
If the answer to https://github.com/phetsims/function-builder/issues/62#issuecomment-224624448 is "no or negligible performance difference", then I'll probably roll back https://github.com/phetsims/function-builder/commit/47d2fbe1d168978ec32f92c153beb2e35c519026.

**Comment by phet-steele** [comment]
@pixelzoom looks fixed on my end.

**Comment by pixelzoom** [comment]
Skype chat with @jonathanolson:

[6/8/16, 10:05:12 AM] Chris Malley: For scenery.Image, is there a significant performance difference between setImage( htmlImageElement ) and setImage( canvas.toDataURL() ) ?
[6/8/16, 10:08:29 AM] Jonathan Olson: Might depend on whether you are rendering as SVG or not
[6/8/16, 10:21:28 AM] Chris Malley: SVG
[6/8/16, 10:22:14 AM] Jonathan Olson: Probably depends on the browser. It isn't something I have tested or noticed yet
[6/8/16, 10:22:22 AM] Jonathan Olson: Does it seem slower one way?
[6/8/16, 10:24:32 AM] Chris Malley: No.  This is related to https://github.com/phetsims/function-builder/issues/62#issuecomment-224485926.   I used `this.imageNode.image = card.image;` in a situation where I thought I might eek out a little more performance, but didn't notice any difference. I'm considering returning to `this.imageNode.setImageWidthSize( card.canvas.toDataURL(),....)` in all cases.
[6/8/16, 10:25:27 AM] Jonathan Olson: Ahh. Generating the data URL will have some overhead
[6/8/16, 10:25:28 AM] Chris Malley: Was also considering startup performance... Ie, not creating canvases for the cards until they are removed from the input carousel.
[6/8/16, 10:26:23 AM] Chris Malley: Canvases are now created on demand, when a card is first removed from input carousel. Formerly created when a card (ImageCard) was constructed.
[6/8/16, 10:27:27 AM] Chris Malley: Since the issue that Steele reported appears to be resolved by using setImageWithSize in all case where I need to change a card's image, then perhaps I'll just leave it the way it is (create canvas on demand).
[6/8/16, 10:30:14 AM] Jonathan Olson: Sure, seems like a fine approach if it doesn't lag user motion

... so I'm going to leave ImageCardNode et. al. as is.  

Closing.
